spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -0,0 +1,566 @@
+      versionsWithRemoveFiles.foreach{
+        case version =>
+          if (versionsWithAddFiles.contains(version) && !ignoreChanges) {
+            throw DeltaSharingErrors.deltaSourceIgnoreChangesError(version)
+          } else if (!versionsWithAddFiles.contains(version) && !ignoreDeletes) {
+            throw DeltaSharingErrors.deltaSourceIgnoreDeleteError(version)
+          }
+      }
+    }
+
+    tableFiles.addFiles
+  }
+
+  /**
+   * Get the latest offset when the streaming query starts. It tries to fetch all table files
+   * from the delta sharing server based on the startingVersion, and then return the latest offset
+   * based on the provided read limit.
+   * @param limits - Indicates how much data can be processed by a micro batch.
+   * @return the latest offset.
+   */
+  private def getStartingOffset(
+    limits: Option[AdmissionLimits] = Some(new AdmissionLimits())): Option[Offset] = {
+
+    val (version, isStartingVersion) = getStartingVersion match {
+      case Some(v) => (v, false)
+      case None => (getOrUpdateLatestTableVersion, true)
+    }
+    if (version < 0) {
+      return None
+    }
+
+    getStartingOffsetFromSpecificDeltaVersion(version, isStartingVersion, limits)
+  }
+
+  override def getDefaultReadLimit: ReadLimit = {
+    new AdmissionLimits().toReadLimit
+  }
+
+  override def latestOffset(startOffset: streaming.Offset, limit: ReadLimit): streaming.Offset = {
+    val limits = AdmissionLimits(limit)
+
+    val currentOffset = if (previousOffset == null) {
+      getStartingOffset(limits)
+    } else {
+      getNextOffsetFromPreviousOffset(previousOffset, limits)
+    }
+    logDebug(s"previousOffset -> currentOffset: $previousOffset -> $currentOffset")
+    currentOffset.orNull
+  }
+
+  override def getOffset: Option[Offset] = {
+    throw new UnsupportedOperationException(
+      "latestOffset(Offset, ReadLimit) should be called instead of this method")
+  }
+
+  override def getBatch(startOffsetOption: Option[Offset], end: Offset): DataFrame = {
+    val endOffset = DeltaSharingSourceOffset(tableId, end)
+
+    val (startVersion, startIndex, isStartingVersion, startSourceVersion) = if (
+      startOffsetOption.isEmpty) {
+      getStartingVersion match {
+        case Some(v) =>
+          // startingVersion is provided by the user
+          (v, -1L, false, None)
+
+        case _ =>
+          // startingVersion is NOT provided by the user
+          if (endOffset.isStartingVersion) {
+            // Get all files in this version if endOffset is startingVersion
+            (endOffset.tableVersion, -1L, true, None)
+          } else {
+            assert(
+              endOffset.tableVersion > 0, s"invalid tableVersion in endOffset: $endOffset")
+            // Load from snapshot `endOffset.tableVersion - 1L` if endOffset is not
+            // startingVersion
+            (endOffset.tableVersion - 1L, -1L, true, None)
```

Comment:
Question: 
Why do we load the snapshot from prev version? 
Do we want to use its metadata ?

Reply:
I don't think it's anything related to the metadata. The schema of the latest table version at the time the stream query starts is used for constructing the dataframe.

I copied this logic from DeltaSource, I guess:
It's more of a product behavior decision, when `getBatch` is called and startOffsetOption is not set, and endOffset is not a startingVersion, it means: 1) no startingVersion is provided, 2) the stream query is not starting from a checkpoint 3) it's the first `getBatch` after the stream query started, 4) endOffset is from a previous `latestOffset` call. 5) and `endOffset` is NOT startingVersion. 6) and given 1) in L410, some version is the startingVersion.

Now, it needs to decide where the starting version is, then I guess it's a prod decision to use (endOffset.tableVersion - 1), or maybe we should use the version we used in L410?

@zsxwing Do you have an idea on this? Thanks!

Reply:
Consider adding to the comment that we do this to maintain same behavior as delta.

Reply:
added the comment, will merge this first and still confirm with Ryan later.

Reply:
This hack for the following case:

When we start a streaming query, we load the latest version X - 1. However, when `latestOffset` is called, we get an offset F in version X (new data got written in between)

Now `getBatch` is called with (startOffsetOption=None, end=the above offset). Since this is the first batch, we need to return the table data in version `X - 1` + the data in version X from offset 0 to offset F (Note: we should not return data after offset F). So we use (offset=-1, version = X-1, isStartingVersion=true) to load the table data in version `X - 1`.

Reply:
@linzhou-db Consider adding Ryan's explanation as a comment.

Reply:
Thanks! I just want to clarify and confirm:
This only happens to the first latestOffset and getBatch call of a stream query without `startingVersion`, and the Offset returned by latestOffset and used as `end` in getBatch can only be 1 version away from the startingVersion returned [here](https://github.com/delta-io/delta-sharing/blob/main/spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala#L407), because it is loading snapshot at startingVersion. 

I'll add the comment in another PR.